x/axelarnet/message_handler.go

```diff
@@ -12,22 +12,84 @@ import (
 	"github.com/ethereum/go-ethereum/crypto"
 
 	"github.com/axelarnetwork/axelar-core/utils/events"
+	axelarnet "github.com/axelarnetwork/axelar-core/x/axelarnet/exported"
 	"github.com/axelarnetwork/axelar-core/x/axelarnet/keeper"
 	"github.com/axelarnetwork/axelar-core/x/axelarnet/types"
 	evmtypes "github.com/axelarnetwork/axelar-core/x/evm/types"
 	nexus "github.com/axelarnetwork/axelar-core/x/nexus/exported"
 	"github.com/axelarnetwork/utils/funcs"
 )
 
+// Fee is used to pay relayer for executing cross chain message
+type Fee struct {
+	Amount    string `json:"amount"`
+	Recipient string `json:"recipient"`
+}
+
 // Message is attached in ICS20 packet memo field
 type Message struct {
 	DestinationChain   string `json:"destination_chain"`
 	DestinationAddress string `json:"destination_address"`
 	Payload            []byte `json:"payload"`
 	Type               int64  `json:"type"`
+	Fee                *Fee   `json:"fee"` // Optional
+}
+
+func (f Fee) validate(ctx sdk.Context, n types.Nexus, b types.BankKeeper, token sdk.Coin, msgType nexus.MessageType) error {
+	amt, ok := sdk.NewIntFromString(f.Amount)
+	if !ok || amt.LTE(sdk.ZeroInt()) {
+		return fmt.Errorf("invalid fee amount")
+	}
+
+	afterFee := token.Amount.Sub(amt)
+	switch msgType {
+	case nexus.TypeGeneralMessage:
+		if afterFee.LT(sdk.ZeroInt()) {
+			return fmt.Errorf("fee amount is greater than transfer value")
+		}
+	case nexus.TypeGeneralMessageWithToken:
+		if afterFee.LTE(sdk.ZeroInt()) {
+			return fmt.Errorf("fee amount is greater or equal to transfer value")
+		}
+	default:
+		return fmt.Errorf("unexpected message type for fee")
+	}
+
+	axelar := funcs.MustOk(n.GetChain(ctx, axelarnet.Axelarnet.GetName()))
+	if !n.IsAssetRegistered(ctx, axelar, token.GetDenom()) {
+		return fmt.Errorf("unregistered fee denom %s", token.GetDenom())
+	}
+
+	addr, err := sdk.AccAddressFromBech32(f.Recipient)
+	if err != nil {
+		return err
+	}
+
+	if b.BlockedAddr(addr) {
+		return fmt.Errorf("fee recipient is a blocked address")
+	}
+
+	return nil
 }
 
-func validateMessage(ctx sdk.Context, ibcK keeper.IBCKeeper, n types.Nexus, ibcPath string, msg Message, token keeper.Coin) error {
+// deductFee transfers fee to the recipient, and returns fee amount
+func deductFee(ctx sdk.Context, b types.BankKeeper, fee *Fee, denom string, msgID string) (sdk.Int, error) {
+	if fee == nil {
+		return sdk.ZeroInt(), nil
+	}
+
+	coin := sdk.NewCoin(denom, funcs.MustOk(sdk.NewIntFromString(fee.Amount)))
+	recipient := funcs.Must(sdk.AccAddressFromBech32(fee.Recipient))
+	events.Emit(ctx, &types.FeePaid{
+		MessageID: msgID,
+		Recipient: recipient,
+		Fee:       coin,
+	})
+
+	return coin.Amount, b.SendCoins(ctx, types.MessageSender, recipient, sdk.NewCoins(coin))
```

Comment:
does this work? The fee is transferred from a different chain right? So is the bank keeper going to find it in the account of the sender?

Reply:
Maybe rename `MessageSender` to `AxelarGMPAccount` or something (since it's our special account) so it doesn't get confused the GMP sender?

Reply:
good with `AxelarGMPAccount`, the token sits on AxelarGMPAccount at this point, so the bank transfer fee amount from gmp account to the recipient.